src/inmanta/deploy/tasks.py

```diff
@@ -87,81 +96,55 @@ class PoisonPill(Task):
-            except KeyError:
-                # Stale resource, can simply be dropped.
-                # May occur in rare races between new_version and acquiring the lock we're under here. This race is safe
-                # because of this check, and an intrinsic part of the locking design because it's preferred over wider
-                # locking for performance reasons.
-                return
-        await self.execute_on_resource(scheduler, agent, resource_details)
-
-    @abc.abstractmethod
-    async def execute_on_resource(
-        self, scheduler: "scheduler.ResourceScheduler", agent: str, resource_details: "state.ResourceDetails"
-    ) -> None:
-        pass
+        intent = await task_manager.get_resource_intent(self.resource, for_deploy=True)
+        if intent is None:
+            # Stale resource, can simply be dropped.
+            return
+        version, resource_details = intent
 
+        status = await self.do_deploy(task_manager, agent, version, resource_details)
 
-class Deploy(OnLatestState):
-    async def execute_on_resource(
-        self, scheduler: "scheduler.ResourceScheduler", agent: str, resource_details: "state.ResourceDetails"
-    ) -> None:
-        status = await self.do_deploy(scheduler, agent, resource_details)
-
-        is_success = status == const.ResourceState.deployed
-
-        async with scheduler._scheduler_lock:
-            # refresh resource details for latest model state
-            new_details: Optional[state.ResourceDetails] = scheduler._state.resources.get(self.resource, None)
-            my_state: state.ResourceState | None = scheduler._state.resource_state.get(self.resource, None)
-            if new_details is not None and new_details.attribute_hash == resource_details.attribute_hash:
-                assert my_state is not None
-                if is_success:
-                    my_state.status = state.ResourceStatus.UP_TO_DATE
-                    my_state.deployment_result = state.DeploymentResult.DEPLOYED
-                else:
-                    # FIXME[#8008]: WDB to Sander: do we set status here as well?
-                    my_state.deployment_result = state.DeploymentResult.FAILED
-                scheduler._work.notify_provides(self)
-            # The deploy that finished has become stale (state has changed since the deploy started).
-            # Nothing to report on a stale deploy.
-            # A new deploy for the current model state will have been queued already.
+        is_success: bool = status == const.ResourceState.deployed
+        await task_manager.report_resource_state(
+            resource=self.resource,
+            attribute_hash=resource_details.attribute_hash,
+            status=state.ResourceStatus.UP_TO_DATE if is_success else state.ResourceStatus.HAS_UPDATE,
+            deployment_result=state.DeploymentResult.DEPLOYED if is_success else state.DeploymentResult.FAILED,
+        )
```

Comment:
I wonder if this should be in a finaly block? i.e. where is the hard exception boundary for e.g.  all the remoting? 

If this is not called, due to an exception, the state of the scheduler is toast, if I understand correctly?

Reply:
You're right. I made the change, but I had to move some code around to keep it coherent (it felt a bit strange to handle some exceptions in `do_deploy` and some others outside of it.

Reply:
It is normal, different types of failure are handled differently and at some level, no exceptions are expected, but they still have to be stopped to protect the main loop.

Reply:
Exactly, but by lifting it up a level, the two cases become one and we're no longer dealing with vagueries purely for the sake of protecting the main loop, we now have the context to know what we're protecting against.